packages/replayio/src/utils/recordings/upload/uploadSourceMaps.ts

```diff
@@ -0,0 +1,91 @@
+import fs from "fs/promises";
+import promiseMap from "p-map";
+import { hashValue } from "../../hashValue";
+import ProtocolClient from "../../protocol/ProtocolClient";
+import { addOriginalSource } from "../../protocol/api/addOriginalSource";
+import { addSourceMap } from "../../protocol/api/addSourceMap";
+import { checkIfResourceExists } from "../../protocol/api/checkIfResourceExists";
+import { createResource } from "../../protocol/api/createResource";
+import { getResourceToken } from "../../protocol/api/getResourceToken";
+import { debug } from "../debug";
+import { LocalRecording } from "../types";
+
+async function ensureResource(client: ProtocolClient, content: string) {
+  const { token } = await getResourceToken(client, { hash: `sha256:${hashValue(content)}` });
+  const resource = {
+    token,
+    saltedHash: `sha256:${hashValue(token + content)}`,
+  };
```

Comment:
I wonder if it would be worth us adding a stronger type for [`Resource.FileHash`](https://static.replay.io/protocol/tot/Resource/#type-FileHash)?

```ts
// https://static.replay.io/protocol/tot/Resource/#type-FileHash
class FileHash {
  content: string;
  token: string;

  constructor(content: string, token: string) {
    this.content = content;
    this.token = token;
  }

  toJSON() {
    const { content, token } = this;

    return `sha256:${hashValue(token + content)}`;
  }
}
```

Maybe overkill.

Reply:
I don't mind the suggestion but both of those calls are (right now) contained in the same function so I don't see it as a problem. It's slightly more clear to me what happens with the optional `token` in the current version since it's all local and without indirection

Reply:
That's fair, though we're going to introducing another instance of this for the `upload-source-maps` command.

Reply:
That one isn't implemented using the protocol client - the protocol required a `recordingId`. The old CLI was using `@replayio/sourcemap-upload` to do that and I have a plan to continue using it for now (let me know if you think we should pull that here).

Reply:
Ah, gotcha. Then let's just skip this for now.